### PR TITLE
ORC-777: Make the vectorized row batch size configurable in MR record readers and writers

### PR DESCRIPTION
**Pull request by sv2000**
### What changes were proposed in this pull request?

The changes in this PR create overloaded constructors in MapreduceRecordReader/Writer classes that accept the rowbatch size as an argument. In order to use the overloaded constructor, the caller is expected to override the getRecordWriter() method (or getRecordReader()), as exemplified in the TestMapreduceOrcOutputFormat class. 

### Why are the changes needed?

The change is needed to provide greater configurability of the ORC MR readers/writers.
Current behavior of OrcInput(Output)Format is to create an ORC reader/writer with the default row batch size of 1024. This may need to be overridden in certain scenarios to prevent OOMs during reads/writes.

### How was this patch tested?

Added unit tests in TestOrcOutputFormat and TestMapreduceOrcOutputFormat, where we extend the OrcOutputFormat class to override the getRecordWriter() method. The change to override the row batch size config was also tested by running a MR job that used the snapshot version of the orc-mapreduce library.

### CODE REVIEW

File: java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java
Comment (dongjoon-hyun) [inline java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java:58]: Is there a reason why we use `DEFAULT_SIZE`, a fixed value?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a contribution.

**Comment by sv2000** [comment]
@dongjoon-hyun Thanks for your comment! With regard to "Is there a reason why we use DEFAULT_SIZE, a fixed value?", I use the default value to maintain backward compatibility when orc.row.batch.size is not configured.

**Review by pgaref — COMMENTED** [review]
Not sure if an extra conf is the right way to go here.

- ```orc.row.batch.size``` is pretty generic even though it is only used for MR
-  So far it was the callers responsibility to set the appropriate batch size so I am wondering if the same should happen for MR -- using an overloaded constructor or even JobConf?

**Comment by sv2000** [comment]
@pgaref IIUC, what you are proposing is to have an overloaded constructor in OrcMapreduceRecordReader/Writer that accepts rowBatchSize as an argument (like we do in this PR), but leave it to the caller to override the getRecordWriter() method, which in turn calls the overloaded constructor. This way we don't need to introduce a separate config. LMK if I am missing anything.

**Comment by pgaref** [comment]
> @pgaref IIUC, what you are proposing is to have an overloaded constructor in OrcMapreduceRecordReader/Writer that accepts rowBatchSize as an argument (like we do in this PR), but leave it to the caller to override the getRecordWriter() method, which in turn calls the overloaded constructor. This way we don't need to introduce a separate config. LMK if I am missing anything.

Hey @sv2000  -- yes you are on point. Just want to avoid having an extra configuration here as it could confuse users. 
If we can introduce a new constructor for Reader/Writer and release the change (on 1.6 or 1.7) it could sort out the issue.
Thoughts?

**Comment by sv2000** [comment]
> > @pgaref IIUC, what you are proposing is to have an overloaded constructor in OrcMapreduceRecordReader/Writer that accepts rowBatchSize as an argument (like we do in this PR), but leave it to the caller to override the getRecordWriter() method, which in turn calls the overloaded constructor. This way we don't need to introduce a separate config. LMK if I am missing anything.
> 
> Hey @sv2000 -- yes you are on point. Just want to avoid having an extra configuration here as it could confuse users.
> If we can introduce a new constructor for Reader/Writer and release the change (on 1.6 or 1.7) it could sort out the issue.
> Thoughts?

@pgaref I am ok with the proposed change. One potential downside though would be that every user that needs to configure the row batch size will have to override the getRecordWriter() method. But may be worth the cost to pay to avoid the confusion of introducing a new config.

**Comment by dongjoon-hyun** [comment]
If we reach an agreement, shall we move forward in that way, @sv2000 ?

**Comment by sv2000** [comment]
@pgaref @dongjoon-hyun Addressed comments on this PR. Can you help take a look?

File: java/mapreduce/src/java/org/apache/orc/mapred/OrcMapredRecordReader.java
Comment (dongjoon-hyun) [inline java/mapreduce/src/java/org/apache/orc/mapred/OrcMapredRecordReader.java:72]: Could you keep the original style here? This looks inconsistent with the other existing one at line 64 and with your new code at line 77.

File: java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java
Comment (dongjoon-hyun) [inline java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java:55]: Please add one empty line between constructors.

File: java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java
Comment (dongjoon-hyun) [inline java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java:57]: Shall we match the indentation with line 49?

File: java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java
Comment (dongjoon-hyun) [inline java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java:62]: Let's put `rowBatchSize` in a new line like the following.
```java
Reader fileReader,
Reader.Options options,
int rowBatchSize) throws IOException {
```

**Comment by dongjoon-hyun** [comment]
To @sv2000 . I left a few minor comments about coding styling.
To @pgaref . Could you confirm that all of your concerns are addressed, please?

**Comment by sv2000** [comment]
Thanks @dongjoon-hyun. I addressed your comments. Can you help take a look?

**Comment by pgaref** [comment]
Hey @sv2000  thanks for making the changes -- thats exactly what I had in mind!
Let update PRs description and some of the Mapreduce Tests to use the new constructors -- will merge after that.

**Review by dongjoon-hyun — APPROVED** [review]
Thanks, @sv2000 . +1 from my side. I'll leave this to @pgaref according to his last comments.

**Comment by sv2000** [comment]
@pgaref Added unit tests to use the overloaded constructors and updated the PR description. LMK if the changes look ok.

**Review by pgaref — APPROVED** [review]
LGTM

**Pull request merged**
